Cleaned up Status native bindings  for CreateAccount, CreateChildAccount, RecoverAccount and ValidateNodeConfig

jonprice opened the pull request:
Work for [Issue #440] 

Refactored  lib/library.go to move  business logic from bindings to StatusAPI for 
* CreateAccount
* CreateChildAccount
* RecoverAccount
* ValidateNodeConfig

Added new unit tests in librarytest.go (due to uses of C package) to  test binding code
Moved existing e2e test testCreateChildAccount and testRecoverAccount from lib/utils.go to e2e package. 
Moved existing unit test testValidateNodeConfig  from lib/utils.go to geth/api/api_test.go

jonprice committed:
vscode config for this project

jonprice committed:
vscode config for this project

jonprice committed:
refactored business logic from lib/library.go ValidateNodeConfig,  CreateAccount, CreateChildAccount ,RecoverAccount into the StatusAPI.

jonprice committed:
moved testCreateChildAccount from lib/utils to e2e.
committing with no changes made to give diff record of test changes

jonprice committed:
Fixed TestCreateChildAccount to work in new e2e package

jonprice committed:
fix lint error in test case

jonprice committed:
moved testRecoverAccount from lib/utils to e2e.
committing with no changes made to give diff record of test changes

jonprice committed:
Fixed TestRecoverAccount to work in e2e package

jonprice committed:
Freeing C.CStrings

jonprice committed:
Merge branch 'develop' of https://github.com/jonprice/status-go into develop

jonprice committed:
removing .vscode/settings.json

themue commented on geth/api/api.go:
```diff
@@ -121,21 +124,46 @@ func (api *StatusAPI) CallRPC(inputJSON string) string {
 // sub-account derivations)
-func (api *StatusAPI) CreateAccount(password string) (address, pubKey, mnemonic string, err error) {
-	return api.b.AccountManager().CreateAccount(password)
+func (api *StatusAPI) CreateAccount(password string) common.AccountInfo {
```
While the idea of bundling the account information into one struct may be nice the error at least has to stay outside:

```
CreateAccount(string) (common.AccountInfo, error)
```

The `StatusAPI` is the pure Go API and has to follow Go idioms.

jonprice replied:
Agreed not Go convention. I wasn't sure which way to go as returning multiple values will mean StatusAPI won't be able to be used by gomobile as mentioned in #440 and #391

Assuming refactoring to return separate err, is leaving the populating of the struct (common.AccountInfo.Error) with the error message acceptable, as this also doesn't feel all that Go like, but simplifies the binding side a lot.

themue replied:
You still could fill the `AccountInfo.Error` as you do.

themue commented on geth/api/api.go:
```diff
@@ -121,21 +124,46 @@ func (api *StatusAPI) CallRPC(inputJSON string) string {
 // Otherwise (when parentAddress != ""), child is derived directly from parent.
-func (api *StatusAPI) CreateChildAccount(parentAddress, password string) (address, pubKey string, err error) {
-	return api.b.AccountManager().CreateChildAccount(parentAddress, password)
+func (api *StatusAPI) CreateChildAccount(parentAddress, password string) common.AccountInfo {
```
Regarding error same here.

themue commented on geth/api/api.go:
```diff
@@ -121,21 +124,46 @@ func (api *StatusAPI) CallRPC(inputJSON string) string {
 // Once master key is re-generated, it is inserted into keystore (if not already there).
-func (api *StatusAPI) RecoverAccount(password, mnemonic string) (address, pubKey string, err error) {
-	return api.b.AccountManager().RecoverAccount(password, mnemonic)
+func (api *StatusAPI) RecoverAccount(password, mnemonic string) common.AccountInfo {
```
Dito.

themue commented on geth/api/api_test.go:
```diff
@@ -0,0 +1,125 @@
+
+	"github.com/status-im/status-go/geth/common"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
```
Why not `testify/assert`?

jonprice replied:
can you clarify the issue, is it that both assert and require are been used?

themue replied:
Oh, sorry, just recognised that you used `require.True()` and similar below. I'm more used to use `assert.True()`. So reacted too fast.

themue commented on geth/api/api_test.go:
```diff
@@ -0,0 +1,125 @@
+	}
+}
+
+type StderrCapture struct {
```
Nice type, but add comments.

themue commented on geth/common/types.go:
```diff
@@ -376,10 +376,11 @@ func (e APIError) Error() string {
-	PubKey   string `json:"pubkey"`
-	Mnemonic string `json:"mnemonic"`
-	Error    string `json:"error"`
+	Address    string `json:"address"`
```
As mentioned above: While adding the error as string for marshalling is fine the original error has to be a return value of the function.

themue commented on geth/common/types.go:
```diff
@@ -394,9 +395,10 @@ func (c StopRPCCallError) Error() string {
+	ID         string `json:"id"`
+	Hash       string `json:"hash"`
+	Error      string `json:"error"`
+	ErrorValue error  `json:"-"`
```
Dito.

themue commented on lib/statusapi.go:
```diff
@@ -0,0 +1,46 @@
+	ResetChainData() error
+	ResetChainDataAsync() (<-chan struct{}, error)
+	CallRPC(inputJSON string) string
+	CreateAccount(password string) common.AccountInfo
```
Here`s a good example how typically errors are returned beside other types.

themue reviewed (commented):
Mostly fine, but see breaking Go idioms.

adambabik commented on geth/api/api.go:
```diff
@@ -121,21 +124,46 @@ func (api *StatusAPI) CallRPC(inputJSON string) string {
+func (api *StatusAPI) CreateAccount(password string) common.AccountInfo {
+	address, pubKey, mnemonic, err := api.b.AccountManager().CreateAccount(password)
+	errString := processError(err)
+	return common.AccountInfo{
```
Handling errors in a Go-way would look like this:

A new type that implements `error` and `Marshaler` interfaces are needed.
```go
type AccountError struct {
	err error
}

func NewAccountError(err error) *AccountError {
 	return &AccountError{err}
}

func (e *AccountError) Error() string {
	return fmt.Sprintf("account error: %s", e.err.Error())
}

func (e *AccountError) MarshalJSON() ([]byte, error)  {
 	return json.Marshal(e.err.Error())
}
```

Usage:
```go
return common.AccountInfo{
  // other fields here
  Error: common.NewAccountError(err),
}

data, _ := json.Marshal(&err)
// string(data) equals: { "error": "some error" }
```

jonprice replied:
Would this be instead of returning err as a separate return value as per themue comment above? Or are you suggesting including AccountError in AccountInfo instead of AccountInfo.Error as well as returning err as a separate value?

adambabik replied:
Actually, @themue's idea is very good. I guess we can return an error as a separate value and use it like this:
```go
accountInfo, err := statusAPI.CreateAccount(string)
if err != nil {
  return makeJSONResponse(err)
}

// marshal accountInfo and return it
```

The only difference is that `AccountInfo` won't contain `Error` field. I am not sure if status-react is prepared for this change. I will figure out that and let you know.

adambabik replied:
@jonprice Let's proceed with @themue's idea to return an error as a second argument.

jonprice replied:
Made the change to return error as a second return param, but left the lib/library.go to return the accountInfo with Error value. 

I did some investigation in status-react to see if I could work out how the AccountInfo.Error was handled. I can't see any code that handles the error message, so I'm not sure if Clojure makes the error handling look transparent, or if there isn't any handling of the error case. Either way, I could not determine the impact of changing how the error case was returned.

adambabik commented on geth/api/api.go:
```diff
@@ -240,3 +268,46 @@ func (api *StatusAPI) NotifyUsers(message string, payload fcm.NotificationPayloa
+	return resp
+}
+
+func processError(err error) string {
```
This is not needed because we don't want to print each error to stderr.

jonprice replied:
Outputting to stderr was from existing code in lib/library.go, so removing would change the existing behaviour. Assuming from this comment this is ok?

adambabik replied:
Yes, you can replace it with `log.Error("what happened", "error", err.Error())` like in other bindings. No unit tests are required to verify if that log actually got logged.

adambabik commented on geth/api/api_test.go:
```diff
@@ -0,0 +1,125 @@
+	}
+}
+
+func testValidateNodeConfig(config string, fn func(common.APIDetailedResponse)) {
```
This function is used once. It's better to inline this in `TestValidateNodeConfig`. Just initialize `statusAPI := StatusAPI{}` before loop.

```go
api := NewStatusAPI()
for _, tc := range testCases {
  resp := api.ValidateJSONConfig(tc.Config)
  tc.Callback(resp)
}
```

adambabik commented on geth/api/api_test.go:
```diff
@@ -0,0 +1,125 @@
+	fn(resp)
+}
+
+func Test_processError(t *testing.T) {
```
`processError` is not needed so is this test.

adambabik commented on geth/common/types.go:
```diff
@@ -376,10 +376,11 @@ func (e APIError) Error() string {
+	Address    string `json:"address"`
+	PubKey     string `json:"pubkey"`
+	Mnemonic   string `json:"mnemonic"`
+	Error      string `json:"error"`
```
Explained in a comment above how to get rid of multiple fields to pass error.

adambabik commented on lib/library.go:
```diff
@@ -110,62 +78,24 @@ func CallRPC(inputJSON *C.char) *C.char {
-	}
-	outBytes, _ := json.Marshal(out)
+	accountInfo := statusAPI.CreateAccount(C.GoString(password))
+	outBytes, _ := json.Marshal(accountInfo)
```
Please handle error as above:
```go
resp, err := json.Marshal(accountInfo)
if err != nil {
  return makeJSONResponse(err)
}
return C.CString(string(resp))
```

jonprice replied:
I don't think Marshaling accountInfo can ever return an error, I couldn't find any way to create a unit test that would cause it to return an error, which is why I thought the uses of outBytes, _ := json.Marshal(out) was common in the existing code.

adambabik commented on lib/library.go:
```diff
@@ -110,62 +78,24 @@ func CallRPC(inputJSON *C.char) *C.char {
-	}
-	outBytes, _ := json.Marshal(out)
+	accountInfo := statusAPI.CreateChildAccount(C.GoString(parentAddress), C.GoString(password))
+	outBytes, _ := json.Marshal(accountInfo)
```
The same remark regarding error handling.

adambabik commented on lib/library.go:
```diff
@@ -110,62 +78,24 @@ func CallRPC(inputJSON *C.char) *C.char {
-	}
-	outBytes, _ := json.Marshal(out)
+	accountInfo := statusAPI.RecoverAccount(C.GoString(password), C.GoString(mnemonic))
+	outBytes, _ := json.Marshal(accountInfo)
```
And here.

adambabik commented on lib/librarytest.go:
```diff
@@ -0,0 +1,242 @@
+	pass1 := C.CString("pass1")
+	pass2 := C.CString("pass2")
+	empty := C.CString("")
+	jsonResult := C.CString(`{"address":"add","pubkey":"Pub","mnemonic":"mne","error":""}`)
```
Maybe we can name it like `accountInfo1JSON` and even try to get this value using `json.Marshal()`?

jonprice replied:
will rename. But I think it's better to be explicit on the expected JSON. The test I'm trying to write is given StatusAPI returns a known AccountInfo, the tested function should return this JSON string. Using json.Marshal to generate the expected result means the test is now testing that given StatusAPI return a known AccountInfo, the function returns json that matches how json.Marshal would Marshal that known AccountInfo. This would mean if in the future json.Marshal changes how it marshaled AccountInfo the test would continue to pass, but it could be a breaking change in the API contract for the binding. 

I will also change the assert to JSONEq, that way changes in format and order of the JSON won't break the test.

adambabik commented on lib/librarytest.go:
```diff
@@ -0,0 +1,242 @@
+	empty := C.CString("")
+	jsonResult := C.CString(`{"address":"add","pubkey":"Pub","mnemonic":"mne","error":""}`)
+	jsonResultError := C.CString(`{"address":"","pubkey":"","mnemonic":"","error":"Error Message"}`)
+	defer func() {
```
Is it needed?

jonprice replied:
Assuming you mean the C.free(unsafe.Pointer(config1)). I think they are needed as C.CString are not cleaned up by GC 

Although as this is a short running test it may not matter if the strings are leaked.

adambabik commented on lib/statusapi.go:
```diff
@@ -0,0 +1,46 @@
+	"github.com/status-im/status-go/geth/params"
+)
+
+type StatusAPI interface {
```
A few comments about this:
1. We don't need all methods here, just ones used in `library.go`,
2. It's possible to compose interfaces so we could have:
```go
type StatusAPIAccount interface {
  CreateAccount(password string) common.AccountInfo
  CreateChildAccount(parentAddress, password string) common.AccountInfo
  RecoverAccount(password, mnemonic string) common.AccountInfo
  VerifyAccountPassword(keyStoreDir, address, password string) (*keystore.Key, error)
  SelectAccount(address, password string) error
}

type StatusAPI interface {
  StatusAPIAccount
  // other interfaces...
}
```
it should make it easier to read and maintain.

adambabik commented on e2e/api/api_test.go:
```diff
@@ -186,3 +188,144 @@ func (s *APITestSuite) TestLogoutRemovesCells() {
 	require.Error(err, "Expected that cells was removed")
 }
+
+func (s *APITestSuite) TestCreateChildAccount() bool {
```
Please compare this with `e2e/accounts/accounts_test.go` as a lot of test cases repeats here.

jonprice replied:
Tests are basically the same.  I think the e2e test probably belong on the StatusAPI that way if any additional logic is added to StatusAPI the test will still cover it.  

Should I delete the duplicated test in e2e/accounts as part of this PR? Or remove the new test from api?

adambabik replied:
We should delete these duplicated tests from here (`e2e/`) if they can be run as unit tests but maybe let's leave them for now. There will be too many changes and anything related to accounts is really important.

adambabik commented on e2e/api/api_test.go:
```diff
@@ -186,3 +188,144 @@ func (s *APITestSuite) TestLogoutRemovesCells() {
+	return true
+}
+
+func (s *APITestSuite) TestRecoverAccount() bool {
```
Please compare this with e2e/accounts/accounts_test.go as a lot of test cases repeats here.

adambabik reviewed (changes requested):
Thanks for your contribution! A added some comments but overall it's the right direction.

jonprice committed:
Separated the StatusAPI interface into seperate jailAPI, accountAPI, transactionAPI, nodeAPI and notificationAPI interfaces and removed unused functions. Also made them private as the interfaces are for use in the lib package only.

jonprice committed:
Updated StatusAPI CreateAccount, CreateChildAccount and RecoverAccount to return seperate error value.
The lib/library.go continue to return the accountInfo as received which will contain the error string if StatusAPI has failed.

jonprice committed:
inlined test function

jonprice committed:
Updated var names to clarify.
Switched to assert.JSONEq so format and order changes won't break the test.

adambabik commented on geth/api/api_test.go:
```diff
@@ -0,0 +1,65 @@
+
+	for _, tc := range testCases {
+		t.Logf("TestValidateNodeConfig: %s", tc.Name)
+		statusAPI := StatusAPI{}
```
Not sure if you need to create `statusAPI` each time. Also, please add a comment why you don't use `NewStatusAPI`. I assume it's because you don't want to create `StatusBackend`.

jonprice replied:
Refactored the code to make it clear that a StatusAPI is not needed to validate the configuration.

adambabik commented on lib/main.go:
```diff
@@ -2,7 +2,7 @@ package main
 import "github.com/status-im/status-go/geth/api"
 
-var statusAPI = api.NewStatusAPI()
+var statusAPI libStatusAPI = api.NewStatusAPI()
```
Is specifying a concrete type required here?

jonprice replied:
libStatusAPI is the interface type, without specifying it the type inference would make statusAPI an api.StatusAPI and I wouldn't be able to override it with a mock in the unit test.

adambabik commented on lib/statusapi.go:
```diff
@@ -0,0 +1,62 @@
+package main
+
+// This is a file contains the interface that the package requires
```
Following https://golang.org/doc/effective_go.html#commentary this comment should be above `package main`. You can also call this file `types.go` or `interface.go` to better reflect what is in here.

adambabik reviewed (approved):
Nice! I added a few minor comments.

jonprice committed:
minor changes to tidy up test cases

divan commented:
I was about to comment some considerations on changing methods in `api.go` and realized that maybe right question to as is: "do we need `StatusAPI` abstraction at all"? Seems like it's redundant proxy between `StatusBackend` and `library.go`.

Custom logic related to formatting/serializing/deserializing responses could be done completely in `library.go` and other `StatusAPI` users (`cmd/statusd/debug/commands.go`, for example) can just use StatusBackend.

Removing `StatusAPI` and api.go would simplify code and tests a lot.

@adambabik @themue @tiabc what do you think about it?

themue commented:
As far as I learned the relation between StatusAPI and StatusBackend it absolutely makes sense, yes. Will you add a new issue for it?

tiabc commented:
That's a good idea. It will only require to:
1. Keep a `Sync` and `Async` versions of some functions in `StatusBackend` which is fine comparing to removing the whole unnecessary abstraction level. I think the idea behind `StatusAPI` was to provide a Facade against the whole application which is a good idea but it seems `StatusBackend` already serves that role.
2. Move logging of the recently added notify functions to `StatusBackend` which is also fine.
3. We can also partially finish #254 under the scope of this issue.

adambabik commented:
> Custom logic related to formatting/serializing/deserializing responses could be done completely in library.go

I wonder if it's gonna be readable as we would mix more Go and cgo code. It's harder to test as cgo is not supported by `go test`. I agree that `StatusAPI` does not provide descriptive methods and it operates on strings a lot but we may still consider having a facade between `library.go` and `StatusBackend`, however, that facade should define properly input/output structs so that we have the self-documentary code.

> We can also partially finish #254 under the scope of this issue.

I would do that separately but this issue is closed... Do we need that new binding then?

divan commented:
@adambabik agree. I can see it as `library.go` having two functions for each command — exported one (say, `StartNode() *C.char`, with cgo types), and unexported function, that does actual logic of calling StatusBackend or something else if needed (say, `startNode() error`).

jonprice commented:
what should happen with this pull request, based on the last few comments should I close the request?

adambabik commented:
We closed it for now as it's staying too long in the pipeline, however, it's still not decided what will happen to `StatusAPI` and `StatusBackend`. Probably, `StatusAPI` will be removed and replaced by an interface.
